compiler/rustc_errors/src/diagnostic.rs

```diff
@@ -630,18 +630,27 @@ impl Diagnostic {
         style: SuggestionStyle,
     ) -> &mut Self {
         assert!(!suggestion.is_empty());
-        debug_assert!(
-            !(suggestion.iter().any(|(sp, text)| sp.is_empty() && text.is_empty())),
-            "Span must not be empty and have no suggestion"
-        );
+
+        let parts = suggestion
+            .into_iter()
+            .map(|(span, snippet)| SubstitutionPart { snippet, span })
+            .collect::<Vec<_>>();
+
+        for (index, part_outer) in parts.iter().enumerate() {
+            debug_assert!(
+                !part_outer.span.is_empty() || !part_outer.snippet.is_empty(),
+                "Span must not be empty and have no suggestion",
+            );
+            for part_inner in parts.iter().skip(index + 1) {
+                debug_assert!(
+                    !part_outer.span.overlaps(part_inner.span),
+                    "suggestion must not have overlapping parts:\n{part_outer:?}\n{part_inner:?}",
+                );
+            }
+        }
```

Comment:
After sorting suggestion:

```suggestion
        let mut parts = parts.iter.peekable();
        while let Some(part) = parts.next() {
            debug_assert!(
                !partspan.is_empty() || !part.snippet.is_empty(),
                "Span must not be empty and have no suggestion",
            );
            let Some(next) = parts.peek() else { break; }
            debug_assert!(
                !part.span.overlaps(next.span),
                "suggestion must not have overlapping parts:\n{part:?}\n{next:?}",
            );
        }
```

Reply:
I think this can be written nicer with `parts.array_windows()`. Something like
```rust
debug_assert!(parts.iter().all(|part| !part.span.is_empty() || !part.snipper.is_empty()), "Span must not be empty and have no suggestion");
parts.array_windows(|[a, b]| debug_assert!(!a.span.overlaps(b.span), "suggestion must not have overlapping parts:\n{a:?}\n{b:?}"));
```